org.omg.sysml/src/org/omg/sysml/expressions/functions/HasTypeFunction.java

```diff
@@ -0,0 +1,49 @@
+/*******************************************************************************
+ * SysML 2 Pilot Implementation
+ * Copyright (c) 2022 Model Driven Solutions, Inc.
+ *    
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *  
+ * You should have received a copy of theGNU Lesser General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  
+ * @license LGPL-3.0-or-later <http://spdx.org/licenses/LGPL-3.0-or-later>
+ *  
+ *******************************************************************************/
+package org.omg.sysml.expressions.functions;
+
+import org.eclipse.emf.common.util.EList;
+import org.omg.sysml.expressions.ModelLevelExpressionEvaluator;
+import org.omg.sysml.expressions.util.EvaluationUtil;
+import org.omg.sysml.lang.sysml.Element;
+import org.omg.sysml.lang.sysml.InvocationExpression;
+import org.omg.sysml.lang.sysml.Type;
+
+public class HasTypeFunction extends BaseFunction {
+
+	@Override
+	public String getOperatorName() {
+		return "hastype";
+	}
+	
+	@Override
+	public EList<Element> invoke(InvocationExpression invocation, Element target, ModelLevelExpressionEvaluator evaluator) {
+		Type testedType = EvaluationUtil.getTypeArgument(invocation);
+		if (testedType != null) {
+			EList<Element> values = evaluator.evaluateArgument(invocation, 0, target);
+			if (values != null) {
+				return EvaluationUtil.booleanResult(values.stream().allMatch(value->EvaluationUtil.hasType(invocation, value, testedType)));
```

Comment:
Shouldn't this be `anyMatch`? (Or the `@` have `allMatch` as well?)

Reply:
This is not the implementation of "@" (which is in the "AtFunction" class). The "hastype" operator is like "istype", but matches for only direct typing (e.g., not typing by supertypes of the given type). (This is in the language because it is a legacy capability from UML action semantics.)